src/datachannel.cc

```diff
@@ -220,6 +218,10 @@ void DataChannel::OnStateChange() {
   TRACE_CALL;
   StateEvent* data = new StateEvent(_jingleDataChannel->state());
   QueueEvent(DataChannel::STATE, static_cast<void*>(data));
+  if (_jingleDataChannel->state() == webrtc::DataChannelInterface::kClosed) {
+    _jingleDataChannel->UnregisterObserver();
+    _jingleDataChannel = nullptr;
+  }
```

Comment:
Note to self: needed to move this out of `Run`, because, AFAICT, after `OnStateChange` is invoked with state `kClosed`, the underlying DataChannel will be freed. Later, when we read `state` from the DataChannel in `Run`, we were getting segfaults.